src/observer.rs

```diff
@@ -47,45 +53,143 @@ pub struct Observer {
 
 impl Default for Observer {
     fn default() -> Self {
-        Self::from_persistent_state(PersistentState::default())
+        Observer::new(String::new())
     }
 }
 
 impl Observer {
-    pub fn from_persistent_state(persistent_state: PersistentState) -> Self {
-        Observer {
-            persistent_state,
+    pub fn new(persistent_file: String) -> Self {
+        let mut observer = Observer {
+            persistent_file,
+            persistent_state: PersistentState::default(),
             enumeration_state: None,
             captured_slot_counters: Default::default(),
             slot_clocks: Default::default(),
             node_table_builders: Default::default(),
-        }
+        };
+        observer.read_persistent_state();
+        observer
     }
 
-    pub fn persistent_state(&self) -> &PersistentState {
-        &self.persistent_state
+    // If a persistent state JSON file exists, prefer its contents over the provided
+    // `persistent_state` argument. This allows the observer to restore previously
+    // captured infrastructure information across runs.
+    pub fn read_persistent_state(&mut self) -> () {
+        if self.persistent_file.is_empty() {
+            log::info!("persistent file is not specified, will not keep persistent state");
+            return;
+        }
+        let file_path = PathBuf::from(&self.persistent_file);
+        if !file_path.is_file() {
+            log::info!(
+                "persistent file: {} not found, starting with empty state",
+                self.persistent_file
+            );
+            return;
+        }
+        match File::open(&file_path).and_then(|mut file| {
```

Comment:
This is a classic [TOCTOU](https://en.wikipedia.org/wiki/Time-of-check_to_time-of-use) bug. The race condition itself isn't an important finding here, but the check is completely avoidable: opening the file may succeed, or it may fail indicating that the file does not exist, or it may fail for some other reason. So, `match` those outcomes:

```rust
        match File::open(&file_path) /*…*/ {
            Ok(data) => {
                // …
            },
            Err(e) if e.kind() == io::ErrorKind::NotFound => {
                log::info!(
                    "persistent file: {} not found, starting with empty state",
                    self.persistent_file
                );
            }
            Err(e) => {
                // …
```